nixos/networkd: reimplement useDHCP in a sensible way

lheckemann opened the pull request:
###### Description of changes

cc people I've talked about this with: @fpletz @ma27 @globin @WilliButz 

This seems like a decent default, with the caveat that the wait-for-online service will not succeed. This can be fixed by setting useDHCP on a specific interface, and doing so is still recommended for this reason.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

lheckemann committed:
nixos/networkd: reimplement useDHCP in a sensible way

lheckemann committed:
nixos/tests/networking: add test for global useDHCP

Ma27 reviewed (commented):
Seems reasonable :+1: 

I guess we should also revert the changes in `nixos-generate-config` then, correct?

Also not sure about the state of networkd-docs in NixOS, but if there are any, this should be noted there as well.

Ma27 commented:
> I think it still makes sense to generate config for known interfaces in order to reduce the impact of the wait-online issue.

I haven't used scripted networking in quite a while, but doesn't that hang if an interface from `networking.interfaces` is not available? THis is e.g. an issue if I have plugged in a smartphone for USB tethering, run `nixos-generate-config`, the tethering-eth iface is part of `hardware-configuration.nix`. When I plug it off and reboot, it potentially hangs, correct? See for instance https://github.com/NixOS/nixpkgs/issues/107908.

If we already go down that route, `networking.useDHCP = true;` in `nixos-generate-config` is IMHO the more sensible default. If somebody knows their network well-enough they can still apply some customizations.

lheckemann commented:
The problem with `networking.useDHCP = true` in the general case is that a device with multiple matching interfaces (e.g. a laptop with WiFi and Ethernet, which is fairly common) will never reach an "online" state if only one of the interfaces is used (which is also fairly common), even though it's perfectly usable and "online" from a user's perspective. I'd say that useDHCP should indeed default to true, but we should _also_ generate per-interface config and comment that.

Ma27 commented:
> I'd say that useDHCP should indeed default to true, but we should also generate per-interface config and comment that.

Would be OK with that. Also, my concerns only apply to the use of scripted networking, how `nixos-generate-config` should behave with `networkd` being the default can be discussed when it's time :)

ncfavier commented on nixos/tests/networking.nix:
```diff
@@ -139,6 +139,25 @@ let
+        client.shell_interact()
+        client.succeed("ping -c 1 192.168.1.1")
+        router.succeed("ping -c 1 192.168.1.1")
+        router.succeed("ping -c 1 192.168.1.2")
+        router.succeed("ping -c 1 192.168.1.2")
```
Same line twice?

Ma27 replied:
I think this should be `client.succeed` here @lheckemann .

Ma27 commented:
I'd like to see this in 22.05 as this is an important step towards networkd adoption.

Ma27 committed:
nixos/tests/networking: test connectivity from both client & router

Ma27 committed:
nixos/network-interfaces-systemd: actually set catchall-iface if `useDHCP = true;`

Previously this wasn't done in the `forEach`-expression for
`cfg.interfaces` and thus `networking.useDHCP` didn't have any effect if
no further interface was statically configured.

Ma27 committed:
nixos/networkd: allow RouteMetric for IPv6AcceptRA

According to `systemd.network(5)` of systemd v249 this is a valid
option.

Fixes evaluation of the wildcard network definitions.

Ma27 commented on nixos/modules/tasks/network-interfaces-systemd.nix:
```diff
@@ -92,6 +86,40 @@ in
             };
           };
         });
+        networks."99-ethernet-default-dhcp" = lib.mkIf cfg.useDHCP {
```
This was actually placed in the wrong section: the `99-ethernet-default-dhcp.network` only gets added if `cfg.interfaces != {}` since it's in the `forEach`-expression of `cfg.interfaces`.

I fixed it in the mayflower fork in https://github.com/mayflower/nixpkgs/commit/6d68cdf6d96030e6f7e1a85c48eb57def29177fa

Ma27 commented on nixos/modules/tasks/network-interfaces-systemd.nix:
```diff
@@ -92,6 +86,40 @@ in
+          # of 1024, so that Ethernet is preferred if both are
+          # available.
+          dhcpV4Config.RouteMetric = 1025;
+          ipv6AcceptRAConfig.RouteMetric = 1025;
```
AFAIU we need to whitelist `RouteMetric` here for `IPv6AcceptRA`: https://github.com/mayflower/nixpkgs/commit/bedb4f214d1b6ea6d89ce1b499376e6d4e5a2a95

lheckemann replied:
I'm a bit confused, why did the test pass if this was an issue? :|

Ma27 reviewed (changes requested):
Also, I'd like to see this implemented:
> I'd say that useDHCP should indeed default to true, but we should also generate per-interface config and comment that.

@lheckemann do you want to apply the remaining tasks or do you want me to take over? :)

Ma27 committed:
nixos/tests/networking/dhcpDefault: actually use networkd for client

Ma27 committed:
nixos/doc: add release-note entry for networking.useDHCP change

Ma27 committed:
nixos/nixos-generate-config: use networking.useDHCP by default

Currently we're still using scripted networking by default. A problem
with scripted networking is that having `useDHCP` on potentially
non-existing interfaces (e.g. an ethernet interface for USB tethering)
can cause the boot to hang.

Closes #107908

lheckemann commented:
I think we need to have another look at the test, because in spite of some glaring issues (as you've pointed out), it was passing!?

lheckemann commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -581,17 +581,19 @@ sub multiLineList {
+  # Enables DHCP on each ethernet and wireless interface. In case of scripted networking
+  # (the default) this is the recommended approach. When using systemd-networkd it's
+  # still possible to use this option, but it's recommended to explicitly set it
+  # per-interface using `networking.interfaces.<interface>.useDHCP`.
```
I'd say that the optimal configuration is to have _both_ when using networkd: explicit useDHCP per interface for the ones that should be configured for `network-online.target`, because they're the ones usually used for internet access, _and_ global useDHCP for any extra interfaces that are used occasionally.

Ma27 committed:
nixos/network-interfaces: remove outdated deprecation information

Ma27 committed:
nixos/nixos-generate-config: update comment for useDHCP

vcunat commented:
FYI, I see:
```
dhcpcd[38576]: read_config: /etc/dhcpcd.conf: No such file or directory
```
I think the only relevant option I have is auto-generated:
```
  networking.useDHCP = lib.mkDefault true;
```
(in particular, no networkd)

ncfavier commented on nixos/modules/tasks/network-interfaces-systemd.nix:
```diff
@@ -80,6 +74,42 @@ in
+          # more likely to result in interfaces being configured to
+          # use DHCP when they shouldn't.
+
+          # We set RequiredForOnline to false, because it's fairly
+          # common for such devices to have multiple interfaces and
+          # only one of them to be connected (e.g. a laptop with
+          # ethernet and WiFi interfaces). Maybe one day networkd will
+          # support "any"-style RequiredForOnline...
```
It does! `systemd.network.wait-online.anyInterface = true;`

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/plans-for-the-networkd-infrastructure/23086/3

jfly commented:
> with the caveat that the wait-for-online service will not succeed

What service is this referring to? Is it `systemd-networkd-wait-online.service`?
